### PR TITLE
feat: Use object storage for clickhouse

### PR DESCRIPTION
**Pull request by danielpanzella**

### CODE REVIEW

**Review by zacharyblasczyk — COMMENTED** [review]
> _**Unsolicited feedback** — I was reviewing the paired [wandb/core#45326](https://github.com/wandb/core/pull/45326) (the server-manifest side: `port`→`http-port`, `clickhouseKeeper` sizing, ClickHouse volume cuts) and dug into this PR to ground the coupling. Since I was already in here, here's what I found. Line refs are pinned to `44e046e`._

The architecture is genuinely well done. My overall read is that it's **close, but not quite mergeable as-is**: one hard blocker, one correctness gap, and one operational gap.

## Strengths (credit due)
- **Coherent port split** — [`conn.go` `toURL()`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/conn.go) uses `TCPPort` (native) for the operator's own admin/migration driver; consumers (weave-trace) get `HTTPPort` via the manifest's `http-port` field. Both keys land in the connection secret; the external BYO-CH path matches.
- **Careful object-storage config** — server-wide `s3_main` policy so even `system_*_log` tables land in the bucket; a cache disk wrapping the S3 disk with a documented sort-order dependency; credentials via secret refs / IAM, never plaintext (`objectstorage.go`).
- **The volume cuts are safely coupled** — the S3 read-cache `max_size` is [80% of the local PV](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/spec.go#L149), so core#45326's shrunk PVCs bound the cache with LRU eviction — can't overfill.
- **Correct reconcile ordering** — bucket resolved before ClickHouse (requeues if not ready), Keeper written before ClickHouse, `objStorage == nil` hard-errors. CHK CRD registered + installed + RBAC present. Idempotent `WriteOwnedFields`/`JSONEqual` avoids churn.

## 🔴 Blocker: the `altinity` test package doesn't compile
[`ToClickHouseVendorSpec`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/spec.go#L121) takes 5 params, and both other call sites pass 5 — but [`spec_test.go:61`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/spec_test.go#L61) passes only 4 (missing the trailing `manifest.Manifest{}`). Since Go builds a package's tests as one unit, this kills the whole `altinity` test binary — `objectstorage_test.go`, `spec_test.go`, and `status_test.go` (the keeper-gating tests) **never run**. `go vet ./...internal/.../altinity` fails with *"not enough arguments in call to ToClickHouseVendorSpec"*.

It slipped through because the visible CI on `v2` PRs is only PR-title + Socket checks — there's **no `go build`/`go test` gate**. Recommend fixing the call **and** adding a Go build/test gate on `v2`, or this recurs. (Fittingly, line 61 is the flagship "backs storage… sets a default policy… wires keeper" test.)

## 🟠 Fail-open Keeper readiness gate
A Keeper in `ConditionUnknown` — which is the *normal* state right after the CHK is created but before its `.status.pods` is populated, and also what any transient API error produces — maps via [`inferState_KeeperReportedReadyType`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/status.go#L196) to `UnknownState`, which is in **neither** [`NotReadyStates`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/common/state.go#L12) **nor** [`stateSignificanceOrder`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/status.go#L122). So it doesn't gate: ClickHouse can be summarized `Healthy` with no confirmed Keeper → a consumer starts and `ReplicatedMergeTree` creation can fail — the exact bootstrap hazard the Keeper exists to prevent. Note the inversion: a *missing* CHK gates (`False`→`Pending`), but a *present-but-unprocessed* one doesn't. Suggest gating positively on `Keeper == True`. (The test that would catch this, `status_test.go`, is in the non-compiling package above.)

## 🟠 Fresh-install-only, but unguarded/undocumented
For an existing managed-CH install, the server-wide [`merge_tree/storage_policy` flip](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/objectstorage.go#L165) has no data-migration step (no `MOVE PARTITION`/re-ingest), so pre-existing local MergeTree parts are stranded on the (now-shrunk) PV; the CHI spec overwrite forces a restart under the new policy; and the PV shrink has no `validateClickHouseChanges` admission guard (Redis forbids `storageSize` changes — CH doesn't) and is a k8s no-op anyway (PVCs can't shrink). Either add an admission guard forbidding `storageSize` decrease + a real data-migration step, **or** explicitly gate/document this as fresh-install-only.

## 🟡 Minor / notes
- **No keeper anti-affinity** → "3 keepers" can co-locate on one node (single-node loss = quorum loss); the `default` sizing tier is `replicas: 1` (SPOF). Consider keeper-scoped pod anti-affinity / `PodDistribution`.
- [`resource.MustParse(spec.StorageSize)`](https://github.com/wandb/operator/blob/44e046ef05c21739d5d29f1f44948c62bbf96043/internal/controller/infra/managed/clickhouse/altinity/spec.go#L146) panics on empty/invalid; the keeper path uses `ParseQuantity` (returns an error). Worth harmonizing.
- **External BYO-CH schema break** — `externalClickhouse.port` → `tcpPort` + `httpPort` (now needs both). Fine pre-GA; worth a release note.
- **No code floor on keeper replicas** — the webhook rejects only *even, non-zero*; manifest-injected sizing bypasses the webhook entirely. A `replicas < 1 || even` check in `ToKeeperVendorSpec` would close it.
- **Possible scope creep:** `reconcile_v2.go` drops `runMigrations`' `anyFailed → return error` — failed migration jobs now only surface in status + a tight 5s requeue (no reconcile error, no job recreation). Unrelated to object storage; flagging for a maintainer's eyes.
- **Coverage** (beyond the compile blocker): the `http-port`/`tcp-port` resolver in `reconciler/pods.go` (app-facing, and a breaking secret-key rename) is untested; the new env-dedup loop is untested; `ResolveObjectStorage`/credentials and `conn.go toURL()` are untested.

## Coupling with core#45326
core#45326 is correct as a diff and I approved it there, but the two are a hard lockstep pair — that manifest can't deploy against an operator without this PR (and vice-versa: this removes the `port` field the old manifest uses). So please don't merge the manifest ahead of this, and ideally land the blocker fix first.

Thanks for the careful work here — the storage/cache/policy modeling in particular is nice. Happy to be wrong on any of the above; I don't have the full v2 context.

File: config/crd/bases/apps.wandb.com_weightsandbiases.yaml [hunk 1]
```diff
@@ -1077,10 +1090,61 @@ spec:
...
+                                    type: array
+                                    x-kubernetes-list-map-keys:
+                                    - name
+                                    x-kubernetes-list-type: map
+                                  limits:
+                                    additionalProperties:
+                                      anyOf:
+                                      - type: integer
+                                      - type: string
+                                      pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
```
Comment (wnevis-cmyk) [inline config/crd/bases/apps.wandb.com_weightsandbiases.yaml:1118]: Goofy regex pattern -- it did the same thing for the PVCs on the bufstream deployment. Seems like it's validating what is probably already being validated downstream, but whatever.

File: internal/controller/infra/managed/clickhouse/altinity/objectstorage.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	s3CacheDiskName = "s3_disk_cache"
+	s3MetadataPath  = "/var/lib/clickhouse/disks/s3_disk/"
+	s3CachePath     = "/var/lib/clickhouse/disks/s3_disk_cache/"
+
+	// storageConfigKey renders to the <storage_configuration> config section.
+	storageConfigKey = "storage_configuration"
+)
+
+// ObjectStorageConn holds the resolved bucket connection used to configure the
+// S3-backed disk: endpoint/region as literals, credentials as secret references.
```
Comment (wnevis-cmyk) [inline internal/controller/infra/managed/clickhouse/altinity/objectstorage.go:37]: Does this support multi-provider, or should I take a pass at it?

File: internal/controller/infra/managed/clickhouse/altinity/objectstorage.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	if bucket == "" {
+		return nil, fmt.Errorf("object store connection secret %q is missing Bucket", secretName)
+	}
+
+	endpoint, err := buildEndpoint(get("Scheme"), get("Host"), get("Port"), bucket, get("Region"), objectStoragePrefix(spec), spec.ObjectStorage.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ObjectStorageConn{
```
Comment (wnevis-cmyk) [inline internal/controller/infra/managed/clickhouse/altinity/objectstorage.go:87]: I think there's a similar function at internal/controller/infra/external/objectstore/connection.go. Maybe we can reuse that? It would probably simply add multi-provider external storage

File: Tiltfile [hunk 4]
```diff
@@ -211,31 +211,6 @@ def operator_dockerfile():
     if settings.get("manifestSource") == "local":
         lines.append("ADD %s /server-manifest" % settings.get("localManifestPath"))
 
-    if settings.get("openshiftSCC"):
```
Comment (wnevis-cmyk) [inline Tiltfile:214]: I think this is fine -- there are a bunch of openshift changes I need to make anyway since we've changed a ton since I originally configured this, so whatever.

**Review by wnevis-cmyk — APPROVED** [review]
